ugvc/pipelines/correct_genotypes_by_imputation.py

```diff
@@ -0,0 +1,477 @@
+    gr_ar = np.full([int((num_alt + 2) * (num_alt + 1) / 2), 2], fill_value=-1, dtype=int)
+    if num_alt == 1:
+        gr_ar = np.array([[0, 0], [0, 1], [1, 1]])
+    elif num_alt == 2:
+        gr_ar = np.array([[0, 0], [0, 1], [1, 1], [0, 2], [1, 2], [2, 2]])
+    else:
+        i = 0
+        for a2 in range(num_alt + 1):
+            for a1 in range(a2 + 1):
+                gr_ar[i, 0] = a1
+                gr_ar[i, 1] = a2
+                i += 1
+    return gr_ar
+
+
+def phred(p) -> np.ndarray:
+    q = -10 * np.log10(np.array(p, dtype=np.float))
+    return q
+
+
+def unphred(q) -> np.ndarray:
+    p = np.power(10, -np.array(q, dtype=np.float) / 10)
+    return p
+
+
+def sort_gt(gt) -> list[int]:
+    if isinstance(gt, tuple):
+        gt_list = list(gt)
+    else:
+        if isinstance(gt, str):
+            gt_list = [int(a) for a in gt.replace("/", "|").split("|")]
+
+    gt_list.sort(key=lambda e: (e is None, e))
+    return tuple(gt_list)
+
+
+def different_gt(gt1, gt2) -> bool:
+    out = sort_gt(gt1) != sort_gt(gt2)
+    return out
+
+
+def _f_along_allele(allele_i, gt_ar, f_hom, f_het):
+    ar = (np.any(gt_ar == allele_i, axis=1) & (gt_ar[:, 0] == gt_ar[:, 1])) * f_hom[allele_i - 1] + (
+        np.any(gt_ar == allele_i, axis=1) & (gt_ar[:, 0] != gt_ar[:, 1])
+    ) * f_het[allele_i - 1]
+    return ar
+
+
+def modify_stats_with_imp(
+    pl_tup, num_alt: int, ds_ar: np.ndarray, current_gt, epsilon: float
+) -> tuple[tuple[int], int, tuple[int]]:
+    """
+    Modifies the PL values bases on the results from the imputation (the ALT dose, DS)
+    Input:
+    pl_tup - A tuple with PLs (as in the vcf)
+    num_alt - number of ALT alleles
+    ds_ar - DS values in numpy array
+    current_gt - The genotype (GT) as in the vcf
+    epsilon - The weight of the imputation in determining the new PL. A number between 0 and 1.
+    Output:
+    The new PLs, The new GQ and the new genotype (GT)
+    """
+    gt_ar = genotype_ordering(num_alt)
+
+    # Modify the PLs with the imputation data
+    # f_het ane f_hom are the probabilities for het or hom, for each allele, capped by the epsilon
+    f_het = np.maximum(epsilon, np.minimum(2 - ds_ar, 1 - epsilon))
+    f_hom = np.minimum(np.maximum(ds_ar, 1) - 1, 1 - epsilon)
+    f_hom = np.maximum(epsilon, f_hom)
+    # "Reshape" the f-values in the form of the gt_ar: For each allele, place the relevant
+    # value (het/hom) in the relevant position.
+    f_allele_ar = np.stack([_f_along_allele(i + 1, gt_ar, f_hom, f_het) for i in range(num_alt)],
+                           axis=1)
+    # The probablity for each genotype is the max across all alleles of that genotype. When there are
+    # missing values (i.e. where there is no imputation data on an allele, use the lowest cap).
+    f_gt_ar = np.amax(np.nan_to_num(f_allele_ar, nan=epsilon), axis=1)
+    # I don't change the PL for the hom-ref genotype, so the f is set to 1
+    f_gt_ar[0] = 1
+
+    pl_unphred = unphred(pl_tup)

```

Comment:
I think I lost you here...
I would need to talk about this in order to understand